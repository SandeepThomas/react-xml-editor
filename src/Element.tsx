import * as React from 'react';

import Attributes from './Attributes';
import Collapsoid from './Collapsoid';
import { IElement } from './Parser';
import TextNode from './TextNode';
import { IActions, IBubbleOptions, IXml } from './types';
import { updateNode } from './Util';

interface IProps {
    actions: IActions;
    attributes?: {[key: string]: string};
    childElements?: IElement[];
    collapsed?: boolean;
    id: string;
    name: string;
    xml: IXml;
}

export default class Element extends React.Component<IProps> {
    private ref: React.RefObject<HTMLDivElement>;

    public constructor(props: IProps) {
        super(props);
        this.onCollapse = this.onCollapse.bind(this);
        this.onClick = this.onClick.bind(this);
        this.ref = React.createRef();
    }

    public render(): React.ReactNode {
        const {
            actions,
            childElements,
            collapsed,
            id,
            name,
            xml,
        } = this.props;

        const elements: React.ReactNode[] = [];
        if (childElements) {
            childElements.forEach((childElement: IElement, index: number) => {
                if (childElement['#name'] !== '__text__') {
                    elements.push(
                        <Element
                            actions={actions}
                            attributes={childElement.$}
                            childElements={childElement.$$}
                            collapsed={childElement['#collapsed']}
                            id={`${id}~\$\$~${index}`}
                            key={index}
                            name={childElement['#name']}
                            xml={xml}
                        />
                    );
                } else if (childElement._) {
                    elements.push(
                        <TextNode
                            actions={actions}
                            id={ `${id}~\$\$~${index}~_` }
                            key={index}
                            text={ childElement._ }
                        />
                    );
                }
            });
        }

        let childrenContainer;
        childrenContainer = (
            <React.Fragment>
                <span className="prominentChildren"/>
                <span className="childrenCollapsed focusable">
                    {childElements && <Collapsoid elements={ childElements }/> }
                </span>
                <div className="children">
                    { elements }
                </div>
            </React.Fragment>
        );

        let elementClass = '';
        if (elements.length === 0) {
            elementClass = ' noChildren';
        }
        if (collapsed) {
            elementClass += ' collapsed';
        }

        const openingTag = (
            <span className="tag opening">
                <span className="punc">&lt;</span>
                <span className="name" onClick={ this.onClick } ref={ this.ref }>{ name }</span>
                { this.getAttributes() }
                <span className="punc slash">/</span>
                <span className="punc">&gt;</span>
            </span>
        );

        let closingTag;
        if (elements.length > 0) {
            closingTag = (
                <span className="tag">
                    <span className="punc">&lt;</span>
                    <span className="punc slash">/</span>
                    <span className="name">{ name }</span>
                    <span className="punc">&gt;</span>
                </span>
            );
        }

        return (
            <React.Fragment>
                <div className={`element${elementClass}`}>
                    <span className="connector">
                        <span className="plusminus" onClick={ this.onCollapse } />
                        <span className="draghandle" draggable="true" />
                    </span>
                    { openingTag }
                    { childrenContainer }
                    { closingTag }
                </div>
            </React.Fragment>
        );
    }

    private onClick() {
        const { actions, id, name } = this.props;
        const bubbleOptions: Partial<IBubbleOptions> = {
            id,
            element: name,
            show: true,
            value: name,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
    }

    private onCollapse() {
        const { actions, collapsed, id, xml } = this.props;
        const collapseId = `${id}~#collapsed`;
        actions.setXml(updateNode(xml, collapseId, !collapsed));
    }

    private getAttributes() {
        const { actions, attributes, id, name } = this.props;
        if (attributes) {
            return (
                <Attributes
                    actions={ actions }
                    attributes={ attributes }
                    element={ name }
                    id={ id }
                />
            );
        }
        return null;
    }
}
