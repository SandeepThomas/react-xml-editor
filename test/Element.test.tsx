import {describe, expect, jest, test} from '@jest/globals';
import * as React from 'react';
import renderer from 'react-test-renderer';

import Element from '../src/Element';

describe('Element component', () => {
    test('expanded element renders properly', () => {
        const component = renderer.create(
            <Element
                actions={{} as any}
                childElements={[{'#name': 'child'}, {'#name': '__text__', _: 'text'}]}
                collapsed={ false }
                name="root"
                id={['root']}
                xml={{ root: { '#name': 'root' } }}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('collapsed element renders properly', () => {
        const component = renderer.create(
            <Element
                actions={{} as any}
                childElements={[{'#name': 'child'}, {'#name': '__text__', _: 'text'}]}
                collapsed
                name="root"
                id={['root']}
                xml={{ root: { '#name': 'root' } }}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('element with attribute renders properly', () => {
        const component = renderer.create(
            <Element
                actions={{} as any}
                attributes={{a: 'b'}}
                collapsed={ false }
                name="root"
                id={['root']}
                xml={{ root: { $: {a: 'b'}, '#name': 'root' } }}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('click collapse button should collapse element and hide bubble', () => {
        const setXml = jest.fn();
        const showBubble = jest.fn();
        const component = renderer.create(
            <Element
                actions={{ setXml, showBubble } as any}
                collapsed={ false }
                name="root"
                id={['root']}
                xml={{ root: { '#name': 'root' } }}
            />
        );
        component.toJSON().children[0].children[0].props.onClick();
        expect(setXml.mock.calls[0][0]).toEqual({
            root: {
                '#collapsed': true,
                '#name': 'root',
            },
        });
        expect(showBubble.mock.calls[0][0]).toEqual({
            show: false,
        });
    });

    test('click open tag should show bubble', () => {
        const showBubble = jest.fn();
        const component = renderer.create(
            <Element
                actions={{ showBubble } as any}
                collapsed={ false }
                name="root"
                id={['root']}
                xml={{ root: { '#name': 'root' } }}
            />,{
                createNodeMock: () => {
                    return {
                        getBoundingClientRect: () => ({
                            left: 88,
                            top: 99,
                        }),
                    };
                }
            }
        );
        const stopPropagation = jest.fn();
        component.toJSON().children[1].children[1].props.onClick({
            stopPropagation,
        });
        expect(showBubble.mock.calls[0][0]).toEqual({
            element: 'root',
            id: ['root'],
            left: 88,
            show: true,
            top: 99,
            value: 'root',
        });
        expect(stopPropagation.mock.calls.length).toBe(1);
    });
});
