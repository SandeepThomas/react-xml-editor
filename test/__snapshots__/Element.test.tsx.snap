// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element component collapsed element renders properly 1`] = `
<div
  className="element collapsed"
>
  <span
    className="connector"
  >
    <span
      className="plusminus"
      onClick={[Function]}
    />
    <span
      className="draghandle"
      draggable="true"
    />
  </span>
  <span
    className="tag opening"
  >
    <span
      className="punc"
    >
      &lt;
    </span>
    <span
      className="name"
      onClick={[Function]}
    >
      root
    </span>
    <span
      className="punc slash"
    >
      /
    </span>
    <span
      className="punc"
    >
      &gt;
    </span>
  </span>
  <span
    className="prominentChildren"
  />
  <span
    className="childrenCollapsed focusable"
  >
    text
  </span>
  <div
    className="children"
  >
    <div
      className="element noChildren"
    >
      <span
        className="connector"
      >
        <span
          className="plusminus"
          onClick={[Function]}
        />
        <span
          className="draghandle"
          draggable="true"
        />
      </span>
      <span
        className="tag opening"
      >
        <span
          className="punc"
        >
          &lt;
        </span>
        <span
          className="name"
          onClick={[Function]}
        >
          child
        </span>
        <span
          className="punc slash"
        >
          /
        </span>
        <span
          className="punc"
        >
          &gt;
        </span>
      </span>
      <span
        className="prominentChildren"
      />
      <span
        className="childrenCollapsed focusable"
      />
      <div
        className="children"
      />
    </div>
    <div
      className="textnode focusable"
      onClick={[Function]}
    >
      <span
        className="value"
      >
        text
      </span>
    </div>
  </div>
  <span
    className="tag closing"
  >
    <span
      className="punc"
    >
      &lt;
    </span>
    <span
      className="punc slash"
    >
      /
    </span>
    <span
      className="name"
    >
      root
    </span>
    <span
      className="punc"
    >
      &gt;
    </span>
  </span>
</div>
`;

exports[`Element component element with attribute renders properly 1`] = `
<div
  className="element noChildren"
>
  <span
    className="connector"
  >
    <span
      className="plusminus"
      onClick={[Function]}
    />
    <span
      className="draghandle"
      draggable="true"
    />
  </span>
  <span
    className="tag opening"
  >
    <span
      className="punc"
    >
      &lt;
    </span>
    <span
      className="name"
      onClick={[Function]}
    >
      root
    </span>
    <span
      className="attributes"
    >
      <span
        className="attribute"
      >
        <span
          className="punc"
        >
           
        </span>
        <span
          className="name"
          onClick={[Function]}
        >
          a
        </span>
        <span
          className="punc"
        >
          =
        </span>
        <span
          className="valueContainer"
          onClick={[Function]}
        >
          <span
            className="punc"
          >
            "
          </span>
          <span
            className="value"
          >
            b
          </span>
          <span
            className="punc"
          >
            "
          </span>
        </span>
      </span>
    </span>
    <span
      className="punc slash"
    >
      /
    </span>
    <span
      className="punc"
    >
      &gt;
    </span>
  </span>
  <span
    className="prominentChildren"
  />
  <span
    className="childrenCollapsed focusable"
  />
  <div
    className="children"
  />
</div>
`;

exports[`Element component expanded element renders properly 1`] = `
<div
  className="element"
>
  <span
    className="connector"
  >
    <span
      className="plusminus"
      onClick={[Function]}
    />
    <span
      className="draghandle"
      draggable="true"
    />
  </span>
  <span
    className="tag opening"
  >
    <span
      className="punc"
    >
      &lt;
    </span>
    <span
      className="name"
      onClick={[Function]}
    >
      root
    </span>
    <span
      className="punc slash"
    >
      /
    </span>
    <span
      className="punc"
    >
      &gt;
    </span>
  </span>
  <span
    className="prominentChildren"
  />
  <span
    className="childrenCollapsed focusable"
  >
    text
  </span>
  <div
    className="children"
  >
    <div
      className="element noChildren"
    >
      <span
        className="connector"
      >
        <span
          className="plusminus"
          onClick={[Function]}
        />
        <span
          className="draghandle"
          draggable="true"
        />
      </span>
      <span
        className="tag opening"
      >
        <span
          className="punc"
        >
          &lt;
        </span>
        <span
          className="name"
          onClick={[Function]}
        >
          child
        </span>
        <span
          className="punc slash"
        >
          /
        </span>
        <span
          className="punc"
        >
          &gt;
        </span>
      </span>
      <span
        className="prominentChildren"
      />
      <span
        className="childrenCollapsed focusable"
      />
      <div
        className="children"
      />
    </div>
    <div
      className="textnode focusable"
      onClick={[Function]}
    >
      <span
        className="value"
      >
        text
      </span>
    </div>
  </div>
  <span
    className="tag closing"
  >
    <span
      className="punc"
    >
      &lt;
    </span>
    <span
      className="punc slash"
    >
      /
    </span>
    <span
      className="name"
    >
      root
    </span>
    <span
      className="punc"
    >
      &gt;
    </span>
  </span>
</div>
`;
